draft-ietf-masque-h3-datagram.md

```diff
@@ -181,74 +181,91 @@ all cases, the maximum permitted value of the H3_DATAGRAM SETTINGS parameter is
+The REGISTER_DATAGRAM_FLOW_ID frame (type=TBD) is used to register a flow ID
+with the QUIC stream that it is sent on. By sending this frame, the sender
+indicates that it will process any received datagram with the corresponding
+flow ID using semantics specific to this frame. For example, if this is sent on
+a client-initiated bidirectional stream that has carried an HTTP request, this
+flow ID will now be interpreted as governed by the method of that request.
 
 ~~~
-  Datagram-Flow-Id = sf-list
+REGISTER_DATAGRAM_FLOW_ID Frame {
+  Type (i) = TBD,
+  Length (i),
+  Flow ID (i),
+  Extension String (..),
+}
 ~~~
+{: #register-frame-format title="REGISTER_DATAGRAM_FLOW_ID HTTP/3 Frame Format"}
 
-The "Datagram-Flow-Id" header field is used to associate one or more datagram
-flows with an HTTP message. As a simple example using a single
-flow, the definition of an HTTP method could instruct the client to use
-its flow ID allocation service to allocate a new flow ID, and
-then the client will add the "Datagram-Flow-Id" header field to its request to
-communicate that value to the server. In this example, the resulting header
-field could look like:
+The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
+The payload consists of:
 
-~~~
-  Datagram-Flow-Id = 2
-~~~
+Flow ID:
 
-List members are flow ID elements, which can be named or unnamed.
-One element in the list is allowed to be unnamed, but all but one elements
-MUST carry a name. The name of an element is encoded in the key of the first
-parameter of that element (parameters are defined in Section 3.1.2 of
-{{STRUCT-FIELD}}). Each name MUST NOT appear more than once in the list. The
-value of the first parameter of each named element (whose corresponding key
-conveys the element name) MUST be of type Boolean and equal to true. The value
-of the first parameter of the unnamed element MUST NOT be of type Boolean. The
-ordering of the list does not carry any semantics. For example, an HTTP method
-that wishes to use four datagram flows for the lifetime of its
-request stream could look like this:
+: The flow ID to associate with the QUIC stream that carried this frame.
+
+Extension String:
+
+: A string of comma-separated key-value pairs to enable extensibility.
+
+The ABNF for the Extension String field is as follows (using syntax from
+{{Section 3.2.6 of RFC7230}}):
 
 ~~~
-  Datagram-Flow-Id = 42, 44; ecn-ect0, 46; ecn-ect1, 48; ecn-ce
+  extension-string = [ ext-member *( "," ext-member ) ]
+  ext-member       = ext-member-key "=" ext-member-value
+  ext-member-key   = token
+  ext-member-value = token
 ~~~
 
-In this example, 42 is the unnamed flow, 44 represents the name
-"ecn-ect0", 46 represents "ecn-ect1", and 48 represents "ecn-ce". Note that,
-since the list ordering does not carry semantics, this example can be
-equivalently encoded as:
+Note that these registrations are unilateral and unidirectional: the sender of
```

Comment:
This feels weird, because this requires a party to make unilateral declarations about peer's flow IDs.  Why not just make the requester associate a flow from its side to a stream?

Reply:
The issue there is that you would then need some sort of acknowledgement to know when your peer has parsed that frame. The unilateral model removes the need for acknowledgements at the HTTP level (for example HTTP/2 has an ACK bit on the SETTINGS frame for this reason).

Reply:
> the sender of the frame unilateraly defines the semantics it will apply to the datagrams it
receives

This feels backwards to me - my inclination is to have the sender define the semantics of datagrams it will send on a flow ID.  This is similar to how unidirectional extension streams work -- the sender-chosen stream type identifies the semantics.

Reply:
I don't think the analogy with stream types is correct. The semantics of stream types are determined by an IANA registry, whereas flow IDs are ephemeral per-connection numbers. If you reverse the direction then things break in the presence of reordering.

Reply:
"I'm listening for X on Flow-ID Y" requires a roundtrip before the sender can even send.  Saying, "I'm sending you X on Flow-ID Y" doesn't break in the face of reordering, you just might lose some datagrams if your stack discards datagrams with unknown Flow IDs.  The receiver can also buffer datagrams on unknown Flow IDs for a while if it's concerned about this case.

Consider my extension for unreliable HTTP upload, where I send a POST request headers on a stream and then the body comes in DATAGRAMs. If I've already seen the SETTINGS from the server and know it supports my extension, why do I need to wait another RTT to acquire a flow-ID?